flutter/dart: remove non-stable versions and bump

thiagokokada opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Closes #108606.

This PR removes non-stable versions of Dart and Flutter. The reason is that Flutter/Dart is really old nowadays, specially the non-stable versions. Having to maintain multiple versions makes bumps more difficult. Removing it and focusing in the stable will make it easier to bump versions (at least I hope).

For those that need to use `beta` or `dev` versions of Dart/Flutter, there is still a way using the following example:

```nix
rec {
  dart-dev = dart.override({
    sources = { "2.12.0-242.0.dev-x86_64-linux" = {...}; }; # supply an attrset with the source so you want to download.
    version = "2.12.0-242.0.dev";
  });
  flutter-dev =  flutterPackages.mkFlutter {
    pname = "flutter-dev";
    version = "1.26.0-12.0.pre";
    src = fetchurl {...}; # supply your own source
    depsSha256 = "..."; # this is the SHA256 of Flutter dependencies, set it to `lib.fakeSha256`, build once and replace with the generated value
    patches = [ ]; # supply your own patches
    dart = dart-dev;
  } {};
}
```

**Bumps:**

dart: 2.10.0 -> 2.10.4, 2.10.4 -> 2.10.5
flutter: 1.22.0 -> 1.22.4, 1.22.4 -> 1.22.5

~~Didn't bump Flutter to version 1.22.5 because it is missing some dependency. I think it is lack of a proper Dart release with this version of Flutter (Google didn't release Dart 2.10.5), so it tries to download the dependency from https://pub.dev/ and fails because we use `--offline` flag during build.~~

~~I don't think this is something that will happen frequently, so I didn't bother trying to fix it, but we may need to create some way to solve dependencies online before building if this is a constantly issue in the future.~~

Flutter derivation is now build in a 2-pass process: the first step download its dependencies from https://pub.dev, creating a local repository cache, and the second step uses the repository cache from the first step to finally build Flutter. This makes the build process much more robust and future proof, since we don't depend on Google packaging Flutter with all dependencies anymore.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thiagokokada committed:
flutter: remove non-stable versions

Related issue #108606.

Nowadays we have multiple outdated versions of Flutter in nixpkgs.
Instead, let's focus in having in having stable versions of Flutter
working.

Users needing to use beta or dev versions of Flutter can use mkFlutter
function, that still exists.

thiagokokada committed:
dart: remove all versions except the one used by flutterPackages.stable

For those that need other versions of Dart, they can use the new
sources parameter instead.

thiagokokada committed:
dart: 2.10.0 -> 2.10.4

thiagokokada committed:
flutter: 1.22.0 -> 1.22.4

thiagokokada commented on pkgs/development/interpreters/dart/default.nix:
```diff
@@ -1,87 +1,34 @@
+, fetchurl
+, unzip
+, version ? "2.10.4"
+, sources ?
```
Adding this here so the user can point to other versions like `beta` or `dev`.

SuperSandro2000 replied:
Is it valid to do ``?`` with null or false or empty string?

thiagokokada replied:
I added some basic assertions, but there is also some checks done later on: https://github.com/NixOS/nixpkgs/blob/31eb0f60571a460dda89ae2c2757a5e43bf78a50/pkgs/development/interpreters/dart/default.nix#L47

thiagokokada commented:
@GrahamcOfBorg build dart flutter

SuperSandro2000 commented on pkgs/development/interpreters/dart/default.nix:
```diff
@@ -1,87 +1,34 @@
+    };
+  }
+}:
 with stdenv.lib;
```
While we are at it: Can we remove this?

thiagokokada replied:
Also removed it from `flutter.nix` file.

SuperSandro2000 commented on pkgs/development/interpreters/dart/default.nix:
```diff
@@ -1,87 +1,34 @@
+{ stdenv
+, fetchurl
+, unzip
+, version ? "2.10.4"
```
Not sure if we want to default to a version. Maybe we should find a different way.

thiagokokada replied:
Why not? I think we want to support only the version of Dart compatible with Flutter.

SuperSandro2000 replied:
I am not sure how common it is to overwrite this but maybe we can remove it entirely. Not sure if it is worth it.

thiagokokada replied:
I think it is worth it to keep it because it is quite common for the Dart is Nixpkgs to not match the Dart version you need it. It is just that almost everyone out there will need a specific version of Dart so we can't support just, like, the 2 or 3 most used versions.

thiagokokada replied:
The idea is, we will support the most recent stable version. If you do need another version, fine, you can call this derivation with the correct parameters and it may work (or not).

SuperSandro2000 commented on pkgs/top-level/aliases.nix:
```diff
@@ -116,6 +116,9 @@ mapAliases ({
   d1x_rebirth = dxx-rebirth; # added 2018-04-25
   d2x_rebirth = dxx-rebirth; # added 2018-04-25
+  dart_dev = throw "Non-stable versions of Dart have been removed."; # added 2020-01-15
+  dart_old = dart; # added 2020-01-15
```
Wouldn't this upgrade the dart version? Not sure if we want this.

thiagokokada replied:
Well, it is either this or a `throw`, because `dart_old` will not work anymore.

SuperSandro2000 replied:
I think a throw would be better to not silently upgrade people.

SuperSandro2000 reviewed (approved):
diff LGTM

thiagokokada commented:
@GRBurst Can we also have your review here?

PeterZainzinger commented:
> Didn't bump Flutter to version 1.22.5 because it is missing some dependency. I think it is lack of a proper Dart release with this version of Flutter (Google didn't release Dart 2.10.5), so it tries to download the dependency from https://pub.dev/ and fails because we use --offline flag during build.

@thiagokokada  I tried your your diff and it works with `1.22.4` but not with `1.22.5` even after bumping dart to `2.10.5` (which is release as by now) with same described error:

 ```
configuring
  no configure script, doing nothing
  building
  Resolving dependencies...
  Because flutter_tools depends on node_preamble any which doesn't exist (could not find package node_preamble in cache), version solving failed.
```

The same issue occurs when trying to use a beta/dev/master version using `mkFlutter`. So I don't think this could be merged as is cause it prevents to use any other version than `1.22.4`

thiagokokada commented:
@PeterZainzinger Huh... going to try, but I don't think this PR is the cause of this issue. If we did the same just by updating how master is current is it would have the same issue.

The fault of this problem is that Flutter or Dart is not shipping with all the dependencies needed to build anymore, so we probably need some changes in how Flutter is packaged.

thiagokokada commented:
If anything, this PR will make easier to fix this issue, because instead of trying to fix it in all channels, now the only channel that matters is stable.

thiagokokada committed:
dart: 2.10.4 -> 2.10.5

PeterZainzinger commented:
Ok, I see your point. I agree that it is outside of the scope of this PR, although I think we should try to fix it anyway is a we make a big change to this module.

I investigated a bit further, downloaded both tarballs (1.22.4 vs 1.22.5) and noticed that the 1.22.5 does not ship node_preamble in its .pub_cache/hosted/pub.dartlang.org directory. The version has not changed at all. 

The workaround is quite obvious, add the pub package as an source and copy it to cache. But that pretty hacky and maybe there are other packages that are missing as well...

PeterZainzinger commented:
I created an issue in the flutter repo. The problem in the tarball is pretty obviously on there side.

https://github.com/flutter/flutter/issues/74589

thiagokokada committed:
flutter: 1.22.4 -> 1.22.5

Needs to do a build in 2-pass now since Google stopped shipping all
dependencies needed to build Flutter. This may be an oversight from them
since they used to ship everything, but this makes the whole build
process more robust.

The first step will download all dependencies from pub, and the
second step will build Flutter. Since we need to build repository
first, we also require a new depsSha256 parameter to be passed, that
represents the SHA256 of the resulting derivation of all Flutter
dependencies downloaded from https://pub.dev.

This commit also makes some changes in mkFlutter, allowing the
user to pass src instead of passing version/channel/filename, allowing
for more flexibility (i.e: building from a local fork of Flutter).

thiagokokada commented:
@PeterZainzinger Actually I do think the issue is in our side. We need to do the build in two steps, the first one will download the deps and the second one will build the derivation. I give it a try in this branch: https://github.com/thiagokokada/nixpkgs/commit/e2e30e0777fe64783109a2a52edda091921d9f3d

It almost works, the dependency resolution now works correctly. But it fails later on with `Unable to open file /build/flutter/bin/cache/flutter_tools.snapshot for writing snapshot`. The old `HOME=../..` trick seems to not work anymore.

If someone knows how to fix this issue it would be welcome.

thiagokokada commented:
Rebased.

thiagokokada commented:
@PeterZainzinger Fixed the issue. Now Flutter 1.22.5 builds and runs (at least on my machine).

thiagokokada commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -30,47 +31,47 @@
       ${dart}/bin/dart --snapshot="$SNAPSHOT_PATH" --packages="$FLUTTER_TOOLS_DIR/.packages" "$SCRIPT_PATH"
       echo "$revision" > "$STAMP_PATH"
       echo -n "${version}" > version
-
-      rm -rf bin/cache/{artifacts,dart-sdk,downloads}
-      rm -f  bin/cache/*.stamp
```
Not really needed since those paths were not in `$out`.

PeterZainzinger commented:
> @PeterZainzinger Fixed the issue. Now Flutter 1.22.5 builds and runs (at least on my machine).

I would like to try.  Not sure if the outputHash in deps.nix is stable. Looks like a bit of an hack to be tbh and something that has to be maintained in the future. But if it works....

thiagokokada commented:
> > @PeterZainzinger Fixed the issue. Now Flutter 1.22.5 builds and runs (at least on my machine).
> 
> I would like to try. Not sure if the outputHash in deps.nix is stable. Looks like a bit of an hack to be tbh and something that has to be maintained in the future. But if it works....

It should be. I took this trick from the official docs: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/maven.section.md#double-invocation

(Yeah, it is for Maven but applied you can apply the idea in other places).

PeterZainzinger commented:
Ok , wasnt aware of that trick. Learned something here, thx! 

In an perfect world there would be something like pub2nix :) But if flutter_tools is something that doesn't evolve with every version I think we are fine here. Maybe a possibility to override this hash from mkFlutter would be nice. 

BTW, Flutter works an my machine as well with 428a6a01cf6da5f7c3b8e999992b7226b98370d3

```
❯ flutter --version                                                                                                                    nix-shell
Flutter 1.22.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 7891006299 (6 weeks ago) • 2020-12-10 11:54:40 -0800
Engine • revision ae90085a84
Tools • Dart 2.10.5
```

thiagokokada commented:
> In an perfect world there would be something like pub2nix :) But if flutter_tools is something that doesn't evolve with every version I think we are fine here.

I think how this derivation is currently it could eventually evolve to a more generic build (like `buildDartPackage`), since it is not that different from other derivations from other languages (like `buildGoPackage`).

>  Maybe a possibility to override this hash from mkFlutter would be nice.

Coincidentally had the same idea. Please test it with commit e3da6eb (everything is the same so it will not even trigger a build).

Can you review and approve @PeterZainzinger. Also need re-reviews from @GRBurst @ericdallo @babariviere and @SuperSandro2000.

thiagokokada committed:
dart/flutter: add thiagokokada as maintainer

thiagokokada commented:
Some last cleanups:
- Renamed `deps.nix` to `repository.nix` (so GitHub don't hide the diffs by default)
- No more dependency on `$HOME` being set, instead use `$PUB_CACHE`
- Add `depsSha256` to `mkFlutter` so we can pass a custom version of Flutter and overwrite its SHA256 from dependencies

PeterZainzinger commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -30,47 +30,44 @@
 let
-  drvName = "flutter-${channel}-${version}";
+  repository = callPackage ./repository.nix {
+    inherit src pname version depsSha256;
```
here dart has to be passed as well. otherwise overriden dart for a custom flutter version does not work

PeterZainzinger replied:
I'm trying to use `dev` with your changes. This requires dart 2.12. So I do:

```nix
        version = "1.26.0-12.0.pre";
        channel = "dev";
        filename = "flutter_linux_${version}-${channel}.tar.xz";
        dartVersion = "2.12.0-133.7";
        dartVersionFull = "2.12.0-133.7.beta";
        mydart = pkgs.dart.override {
          version = dartVersion;
          sources = let
            base = "https://storage.googleapis.com/dart-archive/channels";
            x86_64 = "x64";
          in {
            "${dartVersion}-x86_64-linux" = builtins.fetchurl {
              url =
                "${base}/beta/release/${dartVersionFull}/sdk/dartsdk-linux-${x86_64}-release.zip";
              sha256 =
                "sha256:0lz4p1x99sj39cym702yd0sny5yqvik4hff4x3ji4sawzn8snzpy";
            };

          };
        };

        myflutter = pkgs.flutterPackages.mkFlutter rec {
          inherit version;
          dart = mydart;
          pname = "flutter";
          patches = [ ];
          src = builtins.fetchurl {
            url =
              "https://storage.googleapis.com/flutter_infra/releases/${channel}/linux/${filename}";
            sha256 =
              "sha256:02hldr1hs2h1hhqxh97dc2r2nrymsawrc5l8a78rnp7nfn3sjazs";
          };
          depsSha256 = "0d7vhk6axgqajy2d9ia9lc6awcnz6cc3n04r7hnh7bx4hb0jv0l1";

        };
```

But this fails because of the not passed overriden dart sdk

thiagokokada replied:
Good catch.

thiagokokada replied:
Done, can you test again?

PeterZainzinger replied:
It builds the dev version but the dev version is broken. I think thats a seperate issue and out of the scope of this one. It works fine the not overriden one.

PeterZainzinger commented:
@thiagokokada other than the note above  the diff looks good to me.

SuperSandro2000 commented on pkgs/development/compilers/flutter/default.nix:
```diff
@@ -1,40 +1,25 @@
-    dart = dart_dev;
+    patches = getPatches ./patches;
+    src = fetchurl {
+      url =
+        "https://storage.googleapis.com/flutter_infra/releases/${channel}/linux/${filename}";
```
```suggestion
      url = "https://storage.googleapis.com/flutter_infra/releases/${channel}/linux/${filename}";
```

thiagokokada replied:
This is done by `nixfmt`, but yeah it is ugly. Undoing it.

SuperSandro2000 commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -30,47 +30,44 @@
-      HOME=../.. # required for pub upgrade --offline, ~/.pub-cache
-                 # path is relative otherwise it's replaced by /build/flutter
+      mkdir -p "$out/bin/cache"
+      export SNAPSHOT_PATH="$out/bin/cache/flutter_tools.snapshot"
```
Just a note: Quotes are not really required because $out can't contain any character which needs quoting.

thiagokokada replied:
Yeah, I know, but I did this for consistency.

SuperSandro2000 commented on pkgs/development/compilers/flutter/repository.nix:
```diff
@@ -0,0 +1,23 @@
+    export FLUTTER_ROOT="$(pwd)"
+    export FLUTTER_TOOLS_DIR="$FLUTTER_ROOT/packages/flutter_tools"
+
+    (cd "$FLUTTER_TOOLS_DIR" && ${dart}/bin/pub get)
```
I am not sure if this causes issues in nixpkgs.

```suggestion
    pushd "$FLUTTER_TOOLS_DIR"
    ${dart}/bin/pub get
    popd
```

thiagokokada replied:
Actually in this case we don't need to go back to the original directory, so I did only the `pushd` part. But in the `flutter.nix`  did both since it express the intention better.

thiagokokada commented:
> Is it valid to do ? with null or false or empty string?

I am going to add some assertions.

thiagokokada committed:
dart: add some assertions

thiagokokada commented:
I think this PR is mostly ready to merge, but still would like a review from @GRBurst for the Dart parts of it.

thiagokokada commented on pkgs/development/compilers/flutter/flutter.nix:
```diff
@@ -30,47 +30,46 @@
+      export PUB_CACHE="${repository}"
+
+      pushd "$FLUTTER_TOOLS_DIR"
+      ${dart}/bin/pub get --offline
```
Using `pub get` here since it works and avoids an offline warning.

thiagokokada commented:
Rebased and fixed conflicts.

@SuperSandro2000 can we have this merged? I think it is in a good state, and while @GRBurst didn't respond it yet it concurs that this is a good idea: https://github.com/NixOS/nixpkgs/issues/108606#issuecomment-761189685

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 109477` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>dart</li>
<li>flutter (flutterPackages.stable)</li>
<li>hover</li>
</ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 109477` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>dart</li>
</ul>
</details>
